GEODE-7670: Add Tests for PR clear

jujoramos opened the pull request:
Added distributed tests to verify that the clear operation on
Partitioned Regions works as expected when there are other
concurrent operations happening on the cache (puts, gets, members
added and members removed).

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [X] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [X] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [X] Is your initial contribution a single, squashed commit?

- [X] Does `gradlew build` run cleanly?

- [X] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

jujoramos commented:
The `:geode-redis:distributedTest` task has failed twice while running CI for this PR:

```
> Task :geode-redis:distributedTest

org.apache.geode.redis.RedisDistDUnitTest > classMethod FAILED
    java.lang.AssertionError: Suspicious strings were written to the log during this run.
    Fix the strings or use IgnoredException.addIgnoredException to ignore.
    -----------------------------------------------------------------------
    Found suspect string in log4j at line 5300

    [error 2020/03/26 13:30:55.981 GMT <GeodeRedisServer-WorkerThread-3> tid=98] org.apache.geode.ToDataException: toData failed on DataSerializer with id=0 for class class java.util.HashMap
```

The above can be safely ignored while reviewing the changes as this `PR` only adds new `DistributedTests` for the new `clear()` feature on Partition Regions.

gesterzhou commented on geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java:
```diff
@@ -0,0 +1,525 @@
+
+  static RegionShortcut[] regionTypes() {
+    return new RegionShortcut[] {
+        RegionShortcut.PARTITION,
```
can you add more shortcuts?

jujoramos replied:
There are other tickets aimed at testing concurrent operations with overflow, persistence, etc. Adding more region types should be done as part of the other tickets (that's why I set the `RegionShortcut` to be configurable through parameters), not this one.

gesterzhou commented on geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java:
```diff
@@ -0,0 +1,525 @@
+
+    // Let all VMs continuously execute gets for 10 seconds.
+    List<AsyncInvocation<Void>> asyncInvocationList = Arrays.asList(
+        server1.invokeAsync(() -> executeGets(entries, 10)),
```
after the first clear, all other gets() will not really do anything. You should re-populateRegion again before next clear.

jujoramos replied:
The test is still valid, though, we aim to verify that there are no issues with concurrent `gets` and concurrent `clears`, doesn't matter how many entries (if any) the region has. There is another method within the class that tests concurrent `gets` + `puts` + `clears`.

gesterzhou commented on geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java:
```diff
@@ -0,0 +1,525 @@
+    Instant finishTime = Instant.now().plusSeconds(durationInSeconds);
+
+    while (Instant.now().isBefore(finishTime)) {
+      region.clear();
```
you need to measure how long a clear() took. If it's more than 5 seconds, something must be wrong.

jujoramos replied:
I'm not following you on this one... why only 5 seconds?, and why should the test verify how long the operation actually took?. The method basically executes the `clear` operation for a specific period of time to be able to interleave this operation with other concurrent ones in the cache. Measuring how long the operation took should be part of another ticket (a performance one, probably).

gesterzhou replied:
5 seconds is rvv domination timeout. If clear operation took more than 5 seconds, something must be wrong. No need to measure other operations' time.

jujoramos replied:
@gesterzhou : I still don't understand what these tests have to do with the fact that a `clear` operation should take less than 5 seconds. The ticket is to purely verify that `clear` can be executed while other cache operations are being executed at the same time on the same region, shouldn't we create an extra ticket to test the scenario you're talking about?.

gesterzhou commented on geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java:
```diff
@@ -0,0 +1,525 @@
+    Instant finishTime = Instant.now().plusSeconds(durationInSeconds);
+
+    while (Instant.now().isBefore(finishTime)) {
+      IntStream.range(0, numEntries).forEach(i -> region.put(String.valueOf(i), "Value_" + i));
```
it's better to change the continuous put into continuous putAll.

jujoramos replied:
Any reason for that?. I can certainly do it, just wondering what's the actual reasoning behind the request.

gesterzhou replied:
putAll will have longer window to reproduce synchronization issues

jujoramos replied:
Cool, will add another test to execute multiple `putAll` instead of regular `put` opertations.

agingade commented on geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java:
```diff
@@ -0,0 +1,525 @@
+        accessor.invokeAsync(() -> executeGets(entries, 10)));
+
+    // Clear the region every 2 seconds for 10 seconds.
+    getVM(coordinatorVM.vmNumber).invoke(() -> executeClears(10, 2));
```
2 secs may be long pauses...Since there is continuous cache operation, how about sleeping for few mili secs (say 50)

jujoramos replied:
Changed.

agingade commented on geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java:
```diff
@@ -0,0 +1,525 @@
+        accessor.invokeAsync(() -> executePuts(entries, 15)));
+
+    // Clear the region every 3 seconds for 15 seconds.
+    getVM(coordinatorVM.vmNumber).invoke(() -> executeClears(15, 3));
```
3 secs long sleep.

jujoramos replied:
Changed.

agingade commented on geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java:
```diff
@@ -0,0 +1,525 @@
+    PartitionedRegion region = (PartitionedRegion) cacheRule.getCache().getRegion(REGION_NAME);
+
+    await().untilAsserted(() -> {
+      PartitionedRegionStats prStats = region.getPrStats();
```
This waits for message pertaining to region creation or status change. There is DistributionMessageObserver that can be used to see if any messages are still in progress.

jujoramos replied:
Agreed, the `DistributionMessageObserver` can be used but it's a test hook and I've heard we want to get rid of that eventually, that's why I chose to use another approach (which seems to work fine, so I'm not sure I should to apply the requested change here).

agingade replied:
I won't foresee removal of observers; again there are multiple ways to do it; it is your choice :)

agingade commented on geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java:
```diff
@@ -0,0 +1,525 @@
+  /**
+   * Asserts that the region data is consistent on requested VMs.
+   */
+  private void assertRegionDataConsistency(List<VM> vms) {
```
This is going over the PRs entry/key set getting all the values. This will be consistent where-ever this is called. 
Are we trying to see if the data in one vm is same as in another vm. Or are we trying to see data in primary buckets are same as in secondary buckets.
Fetching the local PR data-set and iterating over it will give the data stored in that vm. 
Or just to see an op could be executed successfully after clear?

jujoramos replied:
The method asserts that the data is consistent across VMs by tacking a snapshot of the region per VM and comparing it against the other running VMs, it doesn't check primaries vs secondaries.

```
    vms.forEach(vm -> vm.invoke(() -> {
      final Map<String, String> thisSnapshot = waitForSilenceAndGetRegionSnapshot();
      assertThat(thisSnapshot).isEqualTo(vm0Snapshot);
      thisSnapshot.forEach((key, value) -> assertThat(value).isEqualTo("Value_" + key));
    }));

```

agingade replied:
What I am trying to says is, since you are iterating over the PR region; it will get the data from all the nodes. Thats remains same across all the nodes.

jujoramos replied:
@agingade : got your point, thanks for catching this. Will change the method to verify that the data both on the primary and secondary buckets are in sync.

agingade commented on geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java:
```diff
@@ -0,0 +1,525 @@
+  /**
+   * The test does the following (the clear coordinator is chosen through parameters):
+   * - Launches threads to continuously execute puts and gets for 60 seconds.
+   * - Clears the Partition Region continuously (once every 10 seconds for 60 seconds).
```
No need to specify time; if someone changes the time and doesn't update it, this will become incorrect.

jujoramos replied:
Done!.

agingade commented on geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java:
```diff
@@ -0,0 +1,525 @@
+   *
+   * TODO: Review once GEODE-7678 / GEODE-7912 are merged (CacheWriter lifecycle might change).
+   */
+  public static class MemberKiller extends CacheWriterAdapter<String, String> {
```
Instead of CacheWriter, you can use DistributionMessageObserver. And can use VM.bounce* to kill vms. No need to wait for CacheWriter messaging implementation.

jujoramos replied:
Same as above, I'm still inclined to use the `CacheWriter` for now instead of the `DistributionMessageObserver` hook. If the lifecycle of the listener changes, I'll be more than happy to use another approach :-). If you think this should block the PR from being merged, though, I'll explore using the `DistributionMessageObserver`.

agingade replied:
I don't think lifecycle of listeners changing, your choice.

agingade commented on geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java:
```diff
@@ -0,0 +1,525 @@
+ * added or removed.
+ */
+@RunWith(JUnitParamsRunner.class)
+public class PartitionedRegionClearWithConcurrentOperationsDUnitTest implements Serializable {
```
This is a good start; other cache ops that can be added in mix are Region ops (destroyRegion, invalidateRegion), querying; Region with index, rebalance, etc.

jujoramos replied:
I'm in favour of having clear names in tests and classes, that's why I prefer to use long ones :-). Regarding the second comment, strongly agreed!.
This PR is to specifically address [GEODE-7670](https://issues.apache.org/jira/browse/GEODE-7670), that's why it only addresses `gets` and `puts`, we can certainly improve the test and add extra operations (or region types) as part of other tickets.

agingade replied:
@jujoramos name change and adding new ops I will leave it to you. The acceptance criteria and ticket heading doesn't match; anyways we could add additional tickets to try out other ops.

jujoramos commented:
@gesterzhou 
I've rebased the `PR` against `feature/GEODE-7665` to include the latest changes from the `clear` feature. There are still some failures while running the tests, concurrent `putAll` with `clear` (test method `clearWithConcurrentPutAllShouldWorkCorrectly`) always reproduces a discrepancy between primary buckets a secondary buckets for `PARTITION_REDUNDANT` regions.
Other tests also sporadically fail when I uncomment the following lines in the `assertRegionBucketsConsistency` method too:
```
assertThat(otherDump.getRvv())
    .as("RegionVersionVector for bucket " + bucketId + " on member " 
        + otherDump.getMember() + " is not consistent with member " + firstDump.getMember())
    .isEqualTo(firstDump.getRvv());
```

jujoramos commented:
@agingade 
According to what we discussed offline about the `clear` operation:
> - The clear may be successful or partial.
> - In case of partial, it will throw `PartitionedRegionPartialClearException`.
> - In both cases, the secondary bucket data should be same as primary (even with member departing).
> - The `RegionVersionVector` between primary and secondary should be same, no matter what.

Having that in mind, I've re-based the `PR` against the latest changes from the `feature/GEODE-7665` branch. There are still some test failures, below are the details:

- Test **clearWithConcurrentPutsShouldWorkCorrectly** always fails for `PARTITION_REDUNDANT` regions, doesn't matter whether the clear coordinator is the actual data host or the accessor node. After the clear finishes, the `RegionVersionVector` is not the same on the primary and secondary copies for a particular bucket.
- Test **clearShouldFailWhenCoordinatorMemberIsBounced** always fails for `PARTITION_REDUNDANT` regions. After the clear finishes, the `RegionVersionVector` is not the same on the primary and secondary copies for a particular bucket.
- Test **clearWithConcurrentPutAllShouldWorkCorrectly** always fails for `PARTITION_REDUNDANT` regions, doesn't matter whether the clear coordinator is the actual data host or the accessor node. After the clear finishes, the `RegionVersionVector` is not the same on the primary and secondary copies for a particular bucket.
- Test **clearWithConcurrentPutsAndGetsShouldWorkCorrectly** always fails for `PARTITION_REDUNDANT` regions, doesn't matter whether the clear coordinator is the actual data host or the accessor node. After the clear finishes, the `RegionVersionVector` is not the same on the primary and secondary copies for a particular bucket.
- Test **clearWithConcurrentPutsShouldWorkCorrectlyWhenNonCoordinatorMembersAreBounced** randomly fails (for `PARTITION_REDUNDANT` regions) with `PartitionedRegionPartialClearException` or inconsistent `RegionVersionVector` across buckets copies, I believe I'm hitting the same scenario as the one @DonalEvans described [here](https://github.com/apache/geode/pull/5095/files#r432000642).

Thoughts?.

agingade commented on geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java:
```diff
@@ -0,0 +1,680 @@
+        for (int j = 1; j < bucketDumps.size(); j++) {
+          BucketDump otherDump = bucketDumps.get(j);
+
+          assertThat(otherDump.getRvv())
```
This validation needs to be changed....The RVV's are maintained as local-rvv (local to the node/member) and array of remote-rvvs (all other members). The RVVs has to be taken for corresponding member from both dumps and compared.

jujoramos commented:
@agingade 

I've updated the `PR` with the changes you suggested (add `bucketRegion.lockRVVForBulkOp()` and `bucketRegion.unlockRVVForBulkOp()` to `PutAllPRMessage.doLocalPutAll`). I've also modified the test to correctly compare the RVVs from the buckets.
I'm still seeing some random failures when comparing (primary vs secondary) the `values` and `versions` stored within the `Bucket`, specially whenever the `clear` is executed concurrently with `putAll`.
That said, I'd suggest to move your suggestions (`bucketRegion.lockRVVForBulkOp()` and `bucketRegion.unlockRVVForBulkOp()`) to a new ticket and only use this `PR` for the actual distributed test.
Please let me know what you think.

gesterzhou commented on geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java:
```diff
@@ -0,0 +1,715 @@
+
+    // Clear the region.
+    getVM(coordinatorVM.vmNumber).invoke(() -> {
+      assertThatThrownBy(() -> cacheRule.getCache().getRegion(REGION_NAME).clear())
```
The PartialClearException could happen here but not 100% will happen. Your assertion could fail here.

jujoramos replied:
@gesterzhou 

There are three members in this test: `ACCESSOR`, `SERVER1` and `SERVER2`, both servers host primary buckets locally for the region. The `clear` operation is executed either from the `ACCESSOR` or `SERVER1`, and `SERVER2` is always killed through the `MemberKiller` class (`CacheWriter.beforeRegionClear()`) before the `clear` is locally executed.
According to what I understand, the `PartitionedRegionPartialClearException` should always be thrown as `SERVER2` doesn't get started (after the initial shutdown) until the `clear` "finishes", meaning that the clear coordinator **_will never be able to get the lock for all the primary buckets_** (the region is configured as `PARTITION`, not as `PARTITION_REDUNDANT`, so there are no secondary buckets that can become primary while `SERVER2` is down).
Am I missing something here?, shouldn't the exception be thrown no matter what, given the above conditions?.

---

As a side note, the test `clearOnNonRedundantPartitionRegionWithConcurrentPutGetRemoveShouldFailWhenNonCoordinatorMembersAreBounced` failed 4 times out of 100 executions, the actual failure depends on whether the `clear` is executed on the `ACCESSOR` or `SERVER1` vm.
The two failures seen when the `clear` is executed on the the `ACCESSOR` node are basically assertion errors, caused by the fact the the `PartitionedRegionPartialClearException` is not thrown (discussed above).
The two failures seen when the `clear` is executed on `SERVER1`, on the other hand, are caused by the fact that the exception message is different from the one expected by the test (I'll change this and update the `PR`).

jujoramos replied:
@gesterzhou 
I've modified the test assertion to always expect the `PartitionedRegionPartialClearException` but don't check the actual message, all the related failures are gone now.
The question still remains, though, the `PartitionedRegionPartialClearException` is always thrown under the conditions set by the test (that's what I would expect), why do you think the mentioned exception shouldn't be thrown 100% of time under these conditions?.

gesterzhou replied:
Only when the primary holder cannot be found any where, the PartitionedRegionPartialClearException will be throw. 

In your case, only killing server2. There're chances and races that the primary buckets are recreated quick enough on server1. So PartitionedRegionPartialClearException is not 100%.

jujoramos replied:
Maybe I'm missing something here, but there's no activity on the cluster at the time `server2` is killed, so the primaries previously hosted by that server shouldn't be recreated anywhere (and there are no secondaries to promote as the region is just `PARTITION`, not `REDUNDANT_PARTITION`).

jujoramos commented:
@agingade @gesterzhou 

After changing the approach used to shutdown the members during the tests (`DistributionMessageObserver` instead of `CacheWriter`) there are no more failures related to the `Timeout during netsearch/netload/netwrite` I was seeing before, CI is fully green so this `PR` is ready to merge once you approve it.

DonalEvans commented on geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java:
```diff
@@ -0,0 +1,733 @@
+  @Rule
+  public CacheRule cacheRule = CacheRule.builder().createCacheInAll().build();
+
+  private VM accessor, server1, server2;
```
It's recommended not to declare multiple variables inline, so this should ideally be split into three separate lines.

DonalEvans commented on geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java:
```diff
@@ -0,0 +1,733 @@
+    }
+  }
+
+  @SuppressWarnings("unused")
```
This warning suppression is not necessary, as the method is used in `coordinatorsAndRegionTypes()` below.

DonalEvans commented on geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java:
```diff
@@ -0,0 +1,733 @@
+  }
+
+  private void initAccessor(RegionShortcut regionShortcut) {
+    @SuppressWarnings("rawtypes")
```
This warning suppression can be removed if the below line is changed to `PartitionAttributes<String, String> attributes = new PartitionAttributesFactory<String, String>()`

DonalEvans commented on geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java:
```diff
@@ -0,0 +1,733 @@
+  }
+
+  private void initDataStore(RegionShortcut regionShortcut) {
+    @SuppressWarnings("rawtypes")
```
This warning suppression can be removed if the below line is changed to `PartitionAttributes<String, String> attributes = new PartitionAttributesFactory<String, String>()`

DonalEvans commented on geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java:
```diff
@@ -0,0 +1,733 @@
+   * @param bucketDump1 First bucketDump.
+   * @param bucketDump2 Second bucketDump.
+   */
+  @SuppressWarnings({"unchecked", "rawtypes"})
```
These warning suppressions can be safely removed if some changes are made in this method, as described below.

DonalEvans commented on geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java:
```diff
@@ -0,0 +1,733 @@
+  @SuppressWarnings({"unchecked", "rawtypes"})
+  private void assertRegionVersionVectorsConsistency(int bucketId, BucketDump bucketDump1,
+      BucketDump bucketDump2) {
+    RegionVersionVector rvv1 = bucketDump1.getRvv();
+    RegionVersionVector rvv2 = bucketDump2.getRvv();
```
These can be changed to `RegionVersionVector<?>` to prevent "rawtypes" warnings.

DonalEvans commented on geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java:
```diff
@@ -0,0 +1,733 @@
+
+    assertThat(rvv1).isNotNull();
+    assertThat(rvv2).isNotNull();
+    Map<VersionSource, RegionVersionHolder> rvv2Members =
+        new HashMap<VersionSource, RegionVersionHolder>(rvv1.getMemberToVersion());
+    Map<VersionSource, RegionVersionHolder> rvv1Members =
+        new HashMap<VersionSource, RegionVersionHolder>(rvv1.getMemberToVersion());
+    for (Map.Entry<VersionSource, RegionVersionHolder> entry : rvv1Members.entrySet()) {
+      VersionSource memberId = entry.getKey();
+      RegionVersionHolder versionHolder1 = entry.getValue();
+      RegionVersionHolder versionHolder2 = rvv2Members.remove(memberId);
```
The following change set will remove the need for suppressed "unchecked" and "rawtypes" warnings in this method:
```suggestion
    Map<VersionSource<?>, RegionVersionHolder<?>> rvv2Members =
        new HashMap<>(rvv1.getMemberToVersion());
    Map<VersionSource<?>, RegionVersionHolder<?>> rvv1Members =
        new HashMap<>(rvv1.getMemberToVersion());
    for (Map.Entry<VersionSource<?>, RegionVersionHolder<?>> entry : rvv1Members.entrySet()) {
      VersionSource<?> memberId = entry.getKey();
      RegionVersionHolder<?> versionHolder1 = entry.getValue();
      RegionVersionHolder<?> versionHolder2 = rvv2Members.remove(memberId);
```

DonalEvans commented on geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java:
```diff
@@ -0,0 +1,733 @@
+   * Continuously execute putAll operations on the PartitionedRegion for the given
+   * durationInSeconds.
+   */
+  private void executePutAlls(final int start, final int finish, final long durationInSeconds) {
```
The arguments "start" and "finish" here might be better named as "startKey" and "endKey" or something similar, to make it clear that they are not related to time.

DonalEvans commented on geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java:
```diff
@@ -0,0 +1,733 @@
+   * Continuously execute removeAll operations on the PartitionedRegion for the given
+   * durationInSeconds.
+   */
+  private void executeRemoveAlls(final int start, final int finish, final long durationInSeconds) {
```
The arguments "start" and "finish" here might be better named as "startKey" and "endKey" or something similar, to make it clear that they are not related to time.

DonalEvans commented on geode-core/src/distributedTest/java/org/apache/geode/internal/cache/PartitionedRegionClearWithConcurrentOperationsDUnitTest.java:
```diff
@@ -0,0 +1,733 @@
+
+    while (Instant.now().isBefore(finishTime)) {
+      region.clear();
+      Thread.sleep(waitTimeInMilliseconds);
```
Instead of a `while` loop with a `Thread.sleep()`, it might be better to use the [Timer](https://docs.oracle.com/javase/7/docs/api/java/util/Timer.html) and [TimerTask](https://docs.oracle.com/javase/7/docs/api/java/util/TimerTask.html) classes to execute clears (specifically the `scheduleAtFixedRate()` method.) Entirely personal preference, but I do know that in general we try not to use `Thread.sleep()` directly in tests.
